File: src/ast/PointerType.java
package ast;

import java.io.StringWriter;

public class PointerType extends BaseNode {
    private BaseNode _child;

    public PointerType(BaseNode child) {
        super(NodeType.PointerType);
        _child = child;
    }

    @Override
    public boolean HasRightPart() {
        return _child.HasRightPart();
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        _child.PrintLeft(writer);
        if (_child.IsArray()) {
            writer.write(" ");
        }

        if (_child.IsArray() || _child.HasFunctions()) {
            writer.write("(");
        }

        writer.write("*");
    }

    @Override
    public void PrintRight(StringWriter writer) {
        if (_child.IsArray() || _child.HasFunctions()) {
            writer.write(")");
        }

        _child.PrintRight(writer);
    }
}

File: src/de/fearlesstobi/demangler/ast/PointerType.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class PointerType(private val child: BaseNode) : BaseNode(NodeType.PointerType) {
    public override fun hasRightPart(): Boolean {
        return child.hasRightPart()
    }

    public override fun printLeft(writer: StringWriter) {
        child.printLeft(writer)
        if (child.isArray) {
            writer.write(" ")
        }
        if (child.isArray || child.hasFunctions()) {
            writer.write("(")
        }
        writer.write("*")
    }

    public override fun printRight(writer: StringWriter) {
        if (child.isArray || child.hasFunctions()) {
            writer.write(")")
        }
        child.printRight(writer)
    }

}